### ISSUE TITLE
[Epic]: Property pane improvements 

### ISSUE THREAD

**Issue by somangshu** [appsmithorg/appsmith#8845]
### Objective

Property Pane Issues listed multiple times which needs UI/UX improvements.
- Large Code Editor
- Keyboard Nav
- Property grouping (organise) and naming convention
- Action selectors

### Success Metrics

| Goal          | Metric         |
| ------------- | -------------- |
| Simplify user experience  | We need to make the editing experience better, overall the user should find it very easy to edit properties for a widget  |

- Navigate across all properties through the keyboard, be able to change values
- Track usage of the expanded code editor compared to the default code editor (based on the minimum number of lines)

### Requirements

- Issue added under the epic need to be solved
- Send events to track usage of the keyboard

### Out of Scope

- Text Editor (low priority)
- property pane controls - ui redesign

### RACI matrix

| | |
| ------------- | ------------- |
| Responsible   |    @aswathkk  |
| Accountable   |     @aswathkk |
| Consulted     | @Nikhil-Nandagopal, @riodeuno, @areyabhishek|
| Informed      |      @somangshu         |

**Comment by somangshu** [comment]
Actionable from the 8th Nov:
- @aswathkk create a plan of action, go through the docs and plan weekly project catchup
- @vasanth-appsmith get back with the designs

**Comment by somangshu** [comment]
Actionable 18th Nov 2021:
- @aswathkk will continue implementation of issues 
- @vasanth-appsmith will finalize the designs
- @aswathkk based on the last designs estimate the rest of the tasks
- We will check all other property pane issues that must be related here @vuiets @somangshu

**Comment by aswathkk** [comment]
Actionable from 3rd Dec, 2021:
- @vasanth-appsmith has to finalize the designs
- @aswathkk will resolve remaining bugs and start with keyboard navigation

**Comment by aswathkk** [comment]
Actionable from 10th Dec, 2021:

- @vasanth-appsmith will re-work on focus state, wrap up the designs for code editing experience by Wednesday and will schedule a design review meeting in between.
- @aswathkk Will resolve the remaining bugs and continue with keyboard navigation.

Keyboard navigation:
- We shouldn't be tab indexing the js toggle option rather we should toggle it by pressing enter key when the property name is focused

Code editing experience:
- While working on it, we have to make sure the changes are not affecting the code editor present in data source page or other fields.

**Comment by aswathkk** [comment]
### Notes from 17th Dec, 2020:

#### Code Editing experience:
- Finalised on having an expandable code editor on the property pane as provided in [this](https://www.figma.com/file/Vb2vkHFKtR2wOrvLx9InDQ/Property-Pane-Improvement?node-id=2%3A262) design
- But, we won’t be doing that for the code editor on the Data sources
- Code editor styles shouldn’t be changed and it should be consistent to that of what we have on the JSEditor
- There should be breadcrumbs on the top section of the expanded code editor indicating the property
- Expanded code editor can be resized horizontally to 100% and that width should be persisted in the browser. But, property pane’s max width should be restricted as we do now.
- When expanded code editor is resized, we need to show scrollbar on the canvas.
- Non-expanded code editor should have a max-width set (may be 12 line height?) and the scroll bar should only available if we hover.
- We will ship this feature within a feature flag for a smoother release

#### Keyboard navigation:
- Focus and active state should be separated
- The tasks need to be broken down and we will ship this feature incrementally

cc: @somangshu, @momcilo-appsmith, @vuiets

**Comment by aswathkk** [comment]
### Update on 31st Dec, 2021:
- Making the action selector keyboard accessible is taking a long time than expected because of the technical difficulties. I hope this should be completed by next week.
- We need to decide on the Tab behaviour of various property controls.

**Comment by aswathkk** [comment]
### Update on 17th Jan, 2021:
- Keyboard navigation experience for various property controls are being documented [here](https://www.notion.so/appsmith/Property-pane-Keyboard-navigation-5cd8d6fce0e0423cbe1fe4f8579e5e5a).
- Some property controls are getting modified in the App Theming project. That needs to be resolved.
- Keyboard navigation in Text Input and JS convertible options need to give a little more thought since the Tab key is being used to insert Tabs as well as to shift focus between property controls.

**Comment by somangshu** [comment]
> * Keyboard navigation in Text Input and JS convertible options need to give a little more thought since the Tab key is being used to insert Tabs as well as to shift focus between property controls.

@momcilo-appsmith can you think about this as well? Let me know if you need a discussion

**Comment by Nikhil-Nandagopal** [comment]
@aswathkk is there a deploy preview we can have a look at?

**Comment by aswathkk** [comment]
@Nikhil-Nandagopal As mentioned earlier, we are working on each controls individually. So, the DPs are different for each of them. The following document contains the DPs
https://www.notion.so/appsmith/Property-pane-Keyboard-navigation-5cd8d6fce0e0423cbe1fe4f8579e5e5a

**Comment by Nikhil-Nandagopal** [comment]
@aswathkk can we have a merged branch where the latest updated changes are there so we can see the progress? Otherwise it's really hard to give feedback and we run the risk of needing to redo things at the end

**Comment by somangshu** [comment]
@aswathkk we could do what Satish had suggested, its `intermediate branch` for bigger features, We do the following:
- Create a PR for a base feature branch: ex - keyboard-navigation-property-pane
- We work on smaller chunks in independent branches for easier code review (like how you are doing right now), the only difference is that we will base this on the new `intermediate branch`
- All code can be QAed / Reviewed on the intermediate branch for us to get all the feedback.
- Once the intermediate branch is ready, We ship everything together.

@aswathkk @Nikhil-Nandagopal do you think it will solve the problem at hand?

The only caveat is that the DP is not built if the base is not release, Ill see how that can be resolved with the help of @mohanarpit

**Comment by aswathkk** [comment]
### Update on 21st Jan
- @aswathkk will discuss with @jsartisan regarding the property controls that are getting modified in Theming project
- @aswathkk and @ankurrsinghal will re-estimate this epic
- @somangshu will be collecting feedback for keyboard navigation behaviour in an input field. That is whether to use `Alt` + `Tab` combination or to just use `Tab`
- @aswathkk will put out at least 2 keyboard navigatable controls to the channel for a general feedback
- As @Nikhil-Nandagopal mentioned, we will be putting out DP in an intermediate branch that contains all the changes.

**Comment by somangshu** [comment]
Note from Aswath:

> We are working on making the entire property pane keyboard accessible. The end goal of this project is to make each and every property controls in the property pane navigable by just using keyboard.
> So far, we’ve decided on the keyboard shortcuts for various controls and they are listed down in the following Notion doc https://www.notion.so/appsmith/Property-pane-Keyboard-navigation-5cd8d6fce0e0423cbe1fe4f8579e5e5a
> We are stuck on how the Keyboard navigation should behave on Text Input fields in property pane.
Normally we use Tab keys to shift the focus from one field to other. But, in case of Text Input fields, Tab key has multiple use cases:
> - Tab is used to enter literal Tabs
> - Tab is used to do autocompletion

> Here are a few solutions that we found on the last meeting:
> Use Option+Tab keys to shift the focus from one field to other. We found that this is working by default in browsers. But, on further investigation, we found that this only works in Chrome and Safari and only on MacOS. In windows or linux, this key combination is used to switch between open applications. So, this solution in not practical.

>Tab should focus the input, but should not make it editable. The input in focus should only be editable when the user press Enter or Space key. Use Esc key to switch from editable mode to focus mode. In that way, pressing Tab while in focus will shift the focus to the next property control.

>Let us know any better keyboard interactions that you can think about. Whatever it is, it should make the user to quickly navigate through all the property controls just using Tab key.


[ref](https://theappsmith.slack.com/archives/C025SE88KNE/p1643014031055100)

**Comment by aswathkk** [comment]
### Update on 4th Feb
- The tasks are re-estimated and this epic is supposed to be completed by March 11th.
- We will be pushing all the finished property controls into release rather than holding them in an intermediate branch. cc: @vuiets, @Nikhil-Nandagopal 
- @aswatthkk will take feedback from the team regarding the keyboard navigation UX for Draggable list control and Option input control and document it.

**Comment by aswathkk** [comment]
### Update on 18th Feb
- @aswath will take a decision on whether to make the close button in the Icon selector’s search box keyboard accessible or not
- @vasanth-appsmith @momcilo-appsmith Think about a platform level design for contextually educating users
- We will not be fully making the date picker control keyboard accessible because of limitations in blueprints. We will make use of some other UI libs instead of Blueprint’s Date picker. But, we will do that later. @ankurrsinghal 
- @vuiets will list down the analytics events that need to be tracked for the keyboard navigation section of this project

**Comment by aswathkk** [comment]
### Update on 25th Feb
- [ ] - We will not be taking up Draggable List & option select controls for development since we need to rethink the best keyboard navigation UX here. cc: @vasanth-appsmith @momcilo-appsmith 
- [ ] - @aswathkk and @arslanhaiderbuttar will be working on making Colorpicker, box shadow & border-radius control keyboard accessible.

**Comment by aswathkk** [comment]
### Update on 4th March

- @aswathkk & @vuiets will document the user flows that are allowed right after selecting a widget in the canvas.
- @aswathkk & @vuiets will document our general philosophy behind the property pane keyboard navigation
- @aswathkk will be making the color picker keyboard accessible. Once the color picker is opened, we will trap the focus inside the color picker — means, pressing tab will cycle through the focusable items inside the colorpicker.
- Revisit icon selector’s keyboard navigation to make it similar to that of color picker’s.
We’ve scoped out larger code editor from this EPIC

**Comment by aswathkk** [comment]
### Update on 11th March
- @vuiets Took a decision on [issue with focusing on widget name on selecting a widget](https://github.com/appsmithorg/appsmith/issues/11199#issuecomment-1066645349)
- We will scope out the keyboard accessibility for JS Mod toggle from this project
- [ ] A decision has to be taken regarding mouse and keyboard interaction for Color and Date pickers. It is being discussed [here](https://theappsmith.slack.com/archives/C025SE88KNE/p1646908219124649).

**Comment by aswathkk** [comment]
### Update on 21st March

- [ ] For JS toggle, we will treat it as a normal button so we can use tab to focus on it.
- [ ] For the Reset property button that got introduced in theming EPIC, we will also make it focusable.
Reset button will be made visible when we focus on the property
- [ ] We will be adding tooltips to all the property name that doesn't have any tooltip

**Comment by vuiets** [comment]
@aswathkk we could remove from the scope of this project and cover it as a part of the `Property Pane Reorganization` Epic
> We will be adding tooltips to all the property name that doesn't have any tooltip

[#4645 | Add tooltips consistently in the Property Pane](https://github.com/appsmithorg/appsmith/issues/4645) is to tackle this.

**Comment by vuiets** [comment]
@aswathkk we will not navigate into property labels and show the tooltip. We'll skip them and make the property controls quickly accessible. This is to stay in line with the objective which is to make it easy to reach a property control, perform a task on it and continue building your app.

cc @momcilo-appsmith @rohitagarwal88

**Comment by vuiets** [comment]
Also @ankurrsinghal did we raise an issue with the keyboard accessibility limitations of Blueprint's Datepicker?

**Comment by aswathkk** [comment]
### Update on 25th March

- [ ] @ankurrsinghal will start a thread about the keyboard highlight state inside the calendar component since @rohitagarwal88 pointed out the design inconsistency there.
- [ ] @ankurrsinghal will document the limitations of Blueprint's date picker component
- [x] @vuiets will take a decision on whether we want to focus on the property name and show the tooltip or not

**Comment by aswathkk** [comment]
The final task pending on this epic is to add an analytics event for keyboard navigation inside the property pane. This is being tracked at #13217 and is in a code review phase right now. Once that is completed, we can close this epic.
@Nikhil-Nandagopal @dilippitchika
